llvm/lib/CheerpUtils/PartialExecuter.cpp

```diff
@@ -0,0 +1,1440 @@
+
+		std::vector<std::pair<const llvm::Value*, BitMask>> knownArgs;
+		uint32_t argIndex=0;
+		for (auto& op : CI.args())
+		{
+			if (argIndex >= numFixedParams)
+				break;	//This is possible when VAArgs are involved, but we are not forwarding those
+
+			if (isValueComputed(op))
+				knownArgs.push_back({calledFunc.getArg(argIndex), getBitMask(op)});
+
+			argIndex++;
+		}
+
+		addStackFrame();
+
+		for (auto& pair : knownArgs)
+			assignToMaps(pair.first, pair.second);
+	}
+
+	bool hasToBeSkipped(llvm::Instruction& I)
+	{
+		if (!areOperandsComputed(I))
+			return true;
+
+		if (isa<CastInst>(I) || isa<BinaryOperator>(I))
+			return false;
+
+		switch (I.getOpcode())
+		{
+			case Instruction::Br:
+			case Instruction::ICmp:
+			case Instruction::FCmp:
+			case Instruction::GetElementPtr:
+			case Instruction::PHI:
+			case Instruction::Ret:
+			case Instruction::Select:
+			case Instruction::Switch:
+			case Instruction::Alloca:
+			{
+				return false;
+			}
+			case Instruction::Call:
+			{
+				const Function* calledFunc = dyn_cast<Function>(cast<CallInst>(I).getCalledFunction());
+
+				if (!calledFunc)
+					return true;
+
+				if (isAlreadyInCallStack(calledFunc))
+					return true;
+
+				if (calledFunc->isDeclaration())
+					return true;
+
+				return false;
+			}
+			case Instruction::Load:
+			{
+				LoadInst& LI = cast<LoadInst>(I);
+
+				GenericValue SRC = getOperandValue(LI.getPointerOperand());
+				GenericValue *Ptr = (GenericValue*)GVTORP(SRC);
+
+				for (const auto& interval : immutableLoadIntervals)
+				{
+					if ((long long)Ptr >= interval.first && (long long)Ptr < interval.second)
+						return false;
+				}
+
+				// TODO(carlo): Possibly even Loads of Alloca might be proven valid
+				break;
+			}
+			default:
+			{
+				//Default case is for Instruction to be skipped
+				return true;
+			}
+		}
+		return true;
+	}
+	void computeValidLoadIntervals(Module& _module)
+	{
+		for (auto& global : _module.globals())
+		{
+			GlobalVariable* GV = dyn_cast<GlobalVariable>(&global);
+			if (!GV)
+				continue;
+			if (!isGlobalVariablePartiallyExecutable(*GV))
+				continue;
+			if (!GV->isConstant())
+				continue;
+
+			// This constraint is here only given that currently the ExecutionEngines assumes GlobalVariable to be named
+			if (!GV->hasName())
+				continue;
+
+			GenericValue SRC = getOperandValue(GV);
+			GenericValue *Ptr = (GenericValue*)GVTORP(SRC);
+			
+			immutableLoadIntervals.push_back({(long long)Ptr, ((long long)Ptr)+getDataLayout().getTypeAllocSize(GV->getInitializer()->getType())});
+		}
+
+		std::sort(immutableLoadIntervals.begin(), immutableLoadIntervals.end());
+	}
+	BasicBlock* findNextBasicBlock(llvm::Instruction& I)
+	{
+		assert(I.isTerminator());
+		assert(!isa<ReturnInst>(I));
+
+		switch (I.getOpcode())
+		{
+			case Instruction::Br:
+			{
+				BranchInst& BI = cast<BranchInst>(I);
+
+				if (BI.isConditional())
+				{
+					if (isValueComputed(BI.getCondition()))
```

Comment:
I am missing something, I remember bitmasks checks here that seemed to make sense before. Have they been moved elsewhere or removed? And if removed, why?

Reply:
Handling of terminators is duplicated in 2 different places, findNextBasicBlock (no checks on known bits since they are done before-hand) AND PartialInterpreter::visitBasicBlock (with checks on known bits). I will unify them, probably by always doing the checks (otherwise passing a parameter no_checks, but seems overly complex)